File: python/ter/core/schema.py
"""Schema definition and validation"""

from typing import Dict, Any, Optional, List
from dataclasses import dataclass, field
import json
from ter.core.types import BaseType, ValidationResult


@dataclass
class VariableDefinition:
    """Definition of an environment variable"""
    name: str
    type: BaseType
    default: Optional[Any] = None
    required: bool = False
    description: Optional[str] = None


class Schema:
    """Environment contract schema"""
    
    def __init__(self):
        self.variables: Dict[str, VariableDefinition] = {}
    
    def define(self, name: str, type_: BaseType, 
               description: Optional[str] = None) -> "Schema":
        """Define a variable"""
        var_def = VariableDefinition(
            name=name,
            type=type_,
            default=type_.get_default(),
            required=type_.is_required(),
            description=description
        )
        self.variables[name] = var_def
        return self
    
    def get_variable(self, name: str) -> Optional[VariableDefinition]:
        """Get variable definition"""
        return self.variables.get(name)
    
    def get_variables(self) -> Dict[str, VariableDefinition]:
        """Get all variables"""
        return self.variables.copy()
    
    def has_variable(self, name: str) -> bool:
        """Check if variable exists"""
        return name in self.variables
    
    def validate(self, values: Dict[str, Any]) -> Dict[str, Any]:
        """Validate values against schema"""
        errors: Dict[str, str] = {}
        validated: Dict[str, Any] = {}
        
        for name, var_def in self.variables.items():
            if name in values:
                result = var_def.type.validate(values[name])
                if result.valid:
                    validated[name] = result.value
                else:
                    errors[name] = result.error
            elif var_def.default is not None:
                validated[name] = var_def.default
            elif var_def.required:
                errors[name] = "Required"
        
        if errors:
            raise ValueError(f"Validation errors: {errors}")
        
        return validated
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert schema to dictionary"""
        variables = {}
        for name, var_def in self.variables.items():
            variables[name] = {
                "type": var_def.type.to_dict(),
                "default": var_def.default,
                "required": var_def.required,
                "description": var_def.description
            }
        
        return {
            "version": "1.0",
            "variables": variables
        }
    
    def to_json(self, indent: int = 2) -> str:
        """Convert schema to JSON"""
        return json.dumps(self.to_dict(), indent=indent)
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "Schema":
        """Create schema from dictionary"""
        from ter.core.types import (
            StringType, NumberType, IntType, FloatType, BooleanType,
            EnumType, URLType, JSONType, SecretType
        )
        
        schema = cls()
        
        for name, var_config in data.get("variables", {}).items():
            type_config = var_config.get("type", {})
            type_name = type_config.get("type")
            
            # Create appropriate type
            if type_name == "string":
                type_obj = StringType()
            elif type_name == "number":
                type_obj = NumberType()
            elif type_name == "int":
                type_obj = IntType()
            elif type_name == "float":
                type_obj = FloatType()
            elif type_name == "boolean":
                type_obj = BooleanType()
            elif type_name == "enum":
                type_obj = EnumType(type_config.get("values", []))
            elif type_name == "url":
                type_obj = URLType()
            elif type_name == "json":
                type_obj = JSONType()
            elif type_name == "secret":
                type_obj = SecretType()
            else:
                continue
            
            # Apply constraints
            if var_config.get("default"):
                type_obj.default(var_config["default"])
            if var_config.get("required"):
                type_obj.mark_required()
            
            schema.define(name, type_obj, var_config.get("description"))
        
        return schema

File: ruby/lib/ter/core/schema.rb
# frozen_string_literal: true

module Ter
  module Core
    # Schema definition for environment variables
    class Schema
      attr_reader :variables

      def initialize
        @variables = {}
      end

      # Define a variable
      def define(name, type)
        raise ArgumentError, 'Name must be a string' unless name.is_a?(String)
        raise ArgumentError, 'Type must be a BaseType' unless type.is_a?(BaseType)

        @variables[name] = type
        type
      end

      # Get a variable definition
      def get_variable(name)
        @variables[name]
      end

      # Check if variable exists
      def has_variable?(name)
        @variables.key?(name)
      end

      # Get all variables
      def get_variables
        @variables.dup
      end

      # Validate all variables against provided values
      def validate(values = {})
        errors = []

        @variables.each do |name, type|
          value = values[name]
          result = type.validate(value)

          unless result[:valid]
            errors << "#{name}: #{result[:error]}"
          end
        end

        { valid: errors.empty?, errors: errors }
      end

      # Export schema to JSON contract
      def to_contract
        contract = {}

        @variables.each do |name, type|
          contract[name] = type_to_contract(type)
        end

        contract
      end

      # Create schema from JSON contract
      def self.from_contract(contract)
        schema = new

        contract.each do |name, type_def|
          type = contract_to_type(type_def)
          schema.define(name, type)
        end

        schema
      end

      private

      def type_to_contract(type)
        contract = { type: type.class.name.split('::').last }

        contract[:required] = type.required if type.required
        contract[:default] = type.default_value unless type.default_value.nil?
        contract[:description] = type.description unless type.description.empty?

        case type
        when Types::IntType
          contract[:min] = type.instance_variable_get(:@min_value) if type.instance_variable_get(:@min_value)
          contract[:max] = type.instance_variable_get(:@max_value) if type.instance_variable_get(:@max_value)
        when Types::NumberType
          contract[:min] = type.instance_variable_get(:@min_value) if type.instance_variable_get(:@min_value)
          contract[:max] = type.instance_variable_get(:@max_value) if type.instance_variable_get(:@max_value)
        when Types::StringType
          contract[:minLength] = type.instance_variable_get(:@min_length) if type.instance_variable_get(:@min_length)
          contract[:maxLength] = type.instance_variable_get(:@max_length) if type.instance_variable_get(:@max_length)
          contract[:pattern] = type.instance_variable_get(:@pattern).source if type.instance_variable_get(:@pattern)
        when Types::EnumType
          contract[:enum] = type.allowed_values
        end

        contract
      end

      def self.contract_to_type(type_def)
        type_name = type_def['type']
        type = case type_name
                when 'StringType'
                  type = Types.string
                  type.min_length(type_def['minLength']) if type_def['minLength']
                  type.max_length(type_def['maxLength']) if type_def['maxLength']
                  type.pattern(Regexp.new(type_def['pattern'])) if type_def['pattern']
                  type
                when 'IntType'
                  type = Types.int
                  type.min(type_def['min']) if type_def['min']
                  type.max(type_def['max']) if type_def['max']
                  type
                when 'NumberType'
                  type = Types.number
                  type.min(type_def['min']) if type_def['min']
                  type.max(type_def['max']) if type_def['max']
                  type
                when 'BooleanType'
                  Types.boolean
                when 'EnumType'
                  Types.enum(*type_def['enum'])
                when 'URLType'
                  Types.url
                when 'JSONType'
                  Types.json
                when 'SecretType'
                  Types.secret
                else
                  raise "Unknown type: #{type_name}"
                end

        type.mark_required if type_def['required']
        type.default(type_def['default']) if type_def['default']
        type.describe(type_def['description']) if type_def['description']

        type
      end
    end
  end
end
